secrets/manager.go

```diff
@@ -0,0 +1,223 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package secrets
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	failedSecretConfigs = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Name: "prometheus_sp_failed_secret_configs",
+			Help: "Current number of secret configurations that failed to load.",
+		},
+	)
+	secretsTotal = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Name: "prometheus_sp_secrets_total",
+			Help: "Current number of secrets.",
+		},
+	)
+)
+
+func yamlSerialize(obj any) ([]byte, error) {
+	if obj == nil {
+		return []byte{}, nil
+	}
+	var buf bytes.Buffer
+	encoder := yaml.NewEncoder(&buf)
+	if err := encoder.Encode(obj); err != nil {
+		return nil, err
+	}
+	if err := encoder.Close(); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+func yamlEqual(x, y any) (bool, error) {
+	yamlX, err := yamlSerialize(x)
+	if err != nil {
+		return false, err
+	}
+	yamlY, err := yamlSerialize(y)
+	if err != nil {
+		return false, err
+	}
+	return bytes.Equal(yamlX, yamlY), nil
+}
+
+type SecretConfig[T any] struct {
+	Name   string `yaml:"name"`
+	Config T      `yaml:"config"`
+}
+
+type SecretEntry[T any] struct {
+	config T
+	secret Secret
+}
+
+type ProviderManager[T any] struct {
+	ctx      context.Context
+	cancelFn func()
+	provider Provider[T]
+	config   Config[T]
+	secrets  map[string]*SecretEntry[T]
+}
+
+func NewProviderManager[T any](ctx context.Context, reg prometheus.Registerer) (ProviderManager[T], error) {
+	m := ProviderManager[T]{
+		ctx:      ctx,
+		cancelFn: func() {},
+		secrets:  make(map[string]*SecretEntry[T]),
+	}
+	if reg != nil {
+		if err := reg.Register(failedSecretConfigs); err != nil {
+			return m, err
+		}
+		if err := reg.Register(secretsTotal); err != nil {
+			return m, err
+		}
+	}
```

Comment:
```suggestion
		reg.MustRegister(failedSecretConfigs, secretsTotal)
```

It's generally ok to panic here, it's usually programmatic error (duplicate metric registered). Will simplify your error handling. Sorry for not mentioning this early